cl/corpus_importer/management/commands/harvard_opinions.py

```diff
@@ -566,57 +635,102 @@ def clean_docket_number(docket_number: str) -> str:
     return docket_number
 
 
-def clean_body_content(case_body: str) -> str:
+def clean_body_content(case_body: str, harvard: bool = False) -> str:
     """Strip all non-alphanumeric characters
 
     :param case_body: Opinion text
+    :param harvard: Are we harvard xml data
     :return:Opinion text with only alphanumeric characters
     """
     soup = BeautifulSoup(case_body, "lxml")
-    return re.sub(r"[^a-zA-Z0-9]", "", soup.text)
+    if not harvard:
+        opinion_text = soup.text
+    else:
+        opinions = []
+        for op in soup.find_all("opinion"):
+            opinions.append(op.text)
+        opinion_text = "".join([op.text for op in soup.find_all("opinion")])
+
+    return re.sub(r"[^a-zA-Z0-9]", "", opinion_text.lower())
 
 
 def match_based_text(
-    case_body: str,
+    harvard_characters: str,
     docket_number: str,
     case_name_full: str,
     possible_cases: List,
+    case_name_abbreviation: str,
+    citation: Citation,
 ) -> Optional[OpinionCluster]:
     """Compare CL text to Harvard content to establish duplicates
 
-    :param case_body: Harvard case body to compare
+    :param harvard_characters: Harvard stripped characters to compare
     :param possible_cases: List of opinions to check against
     :param docket_number: The docket number
     :param case_name_full: The full case name
     :return: OpinionCluster or None
     """
-    harvard_characters = clean_body_content(case_body)
     for case in possible_cases:
         cl_case_body = get_opinion_content(case)
         cl_characters = clean_body_content(cl_case_body)
+        if len(cl_characters) == 0:
+            logger.info(f"Empty Courtlistener opinion cluster: {case.id}")
+            continue
+
         diff = len(harvard_characters) / len(cl_characters)
         if not (0.3 < diff < 3):
             # Content too dissimilar in length to compare
             continue
 
         percent_match = compare_documents(harvard_characters, cl_characters)
-        if percent_match < 45:
+        if percent_match < 60:
             continue
-        if percent_match < 75 and len(harvard_characters) > 500:
-            # Require a name overlap for good but not great matches.
-            overlaps = overlap_case_names(case.case_name, case_name_full)
-            if not overlaps:
-                continue
-        elif percent_match < 98 and len(harvard_characters) < 500:
+
+        # Require some overlapping case title
+        overlaps = overlap_case_names(
+            case.case_name, [case_name_full, case_name_abbreviation]
+        )
+        if not overlaps:
+            continue
+
+        vector_match = get_cosine_similarity(
+            case.case_name, [case_name_full, case_name_abbreviation]
+        )
```

Comment:
Since you only use get_cosine_similarity in this one place, I think instead of having that big `get_cosine_similarity` function that doesn't do much, you should delete it, and import the (identically named) `string_diff.get_cosine_similarity` function, and just use it like so:

```
# Put a comment from that function here...
best_vector_match = max(
    map(
        lambda x: get_cosine_similarity(case.case_name, x), 
        [case_name_full, case_name_abbreviation]
    )
)
```

Or if that's too fancy (it may well be?):

```
# Put a comment from that function here...
best_vector_match_score = max(
    get_cosine_similarity(case.case_name, case_name_full),
    get_cosine_similarity(case.case_name, case_name_abbreviation),
)
```

Well, that second version sure is easier to read, but I keep thinking I'll use `map` more someday...perhaps not today unless you prefer it. Anyway, seems like the whole

Reply:
Hm, I see now that this parallels the signature of overlap_case_names, but I think I still find it to be an improvement.

Reply:
you are exactly right.